### ISSUE TITLE
API for inserting a Span<T> into a List<T> efficiently

### ISSUE THREAD

**Issue by josetr** [dotnet/corefx#36198]
Perhaps an `AddRange` overload taking a `Span<T>` can be added to the `List` class?

```
var list = new List<byte>();
list.AddRange(new Span<byte>());
```

**Workarounds**

```
foreach(var b in span)
    list.Add(b); // Inefficient only copying one byte at a time
```

```
var list = new List<byte>();
var span = new Span<byte>();
list.AddRange(span.ToArray()); Silly allocation
```

**Comment by ahsonkhan** [comment]
Do you have a scenario to help motivate this addition (maybe example code today where the workarounds are noticeably inefficient)? Are there other List APIs that fall into this category (for instance what about the `List<T>` ctor)?

Similarly, we have the CopyTo that takes arrays.

Presumably, there was a reason such APIs didn't meet the bar during the first round:
https://github.com/dotnet/corefx/issues/21281

> There are some other namespaces that could likely benefit from Span/Buffer, but we should probably handle separately from this initial push:
- System.Collections. It’s not clear to me to what extent we should add span/buffer-based APIs to collections.  CopyTo overloads that take `Span<T>`?

**Comment by DaZombieKiller** [comment]
Just to add to the list of workarounds:

```cs
public static class ListExtensions
{
    static readonly IntPtr itemsOffset = Marshal.OffsetOf(typeof(List<>), "_items");
    static readonly IntPtr  sizeOffset = Marshal.OffsetOf(typeof(List<>), "_size");

    public static Span<T> AsSpan<T>(this List<T> self, int start)             => AsSpan(self).Slice(start);
    public static Span<T> AsSpan<T>(this List<T> self, int start, int length) => AsSpan(self).Slice(start, length);

    public static Span<T> AsSpan<T>(this List<T> self)
    {
        var raw = Unsafe.As<List<T>, RawData>(ref self);
        return Unsafe.As<byte, T[]>(ref Unsafe.Add(ref raw.Data, itemsOffset));
    }

    public static void SetCount<T>(this List<T> self, int count)
    {
        var raw = Unsafe.As<List<T>, RawData>(ref self);
        Unsafe.As<byte, int>(ref Unsafe.Add(ref raw.Data, sizeOffset)) = count;
    }

    class RawData
    {
        public byte Data;
    }
}
```

Which would be used as follows:

```cs
var list = new List<int>(4) { 1, 2, 3, 4 };
var span = new[] { 5, 6, 7, 8 }.AsSpan();

if (list.Count + span.Length > list.Capacity)
    list.Capacity = list.Count + span.Length;

span.CopyTo(list.AsSpan(list.Count));
list.SetCount(list.Count + span.Length);

foreach (var value in list) Console.WriteLine(value);
```

It feels *incredibly* dirty to do this though, so a proper solution in CoreFX would be greatly appreciated.

**Comment by jkotas** [comment]
> It feels incredibly dirty to do this though

Yes, it is dirty and it does not even work. You will get `Type 'System.Collections.Generic.List`1[T]' cannot be marshaled as an unmanaged structure; no meaningful size or offset can be computed.` exception thrown from the `Marshal.OffsetOf` call.

**Comment by jkotas** [comment]
The best current workaround for this is to roll your own `List<T>` type that does exactly what you need.

**Comment by DaZombieKiller** [comment]
> 
> 
> > It feels incredibly dirty to do this though
> 
> Yes, it is dirty and it does not even work. You will get `Type 'System.Collections.Generic.List`1[T]' cannot be marshaled as an unmanaged structure; no meaningful size or offset can be computed.`exception thrown from the`Marshal.OffsetOf` call.

Interestingly, it worked perfectly fine while running on Mono (must be an oversight on their part). However on .NET Core I get the error you mentioned. It's fixable by using reflection, although it requires some extra ceremony.

> 
> 
> The best current workaround for this is to roll your own `List<T>` type that does exactly what you need.

I'd normally recommend this too, but it's not an option when you're dealing with a third party API taking a `List<T>`.
